marimo/_ast/codegen.py

```diff
@@ -174,59 +263,110 @@ def generate_filecontents(
     # Let's keep it disabled by default.
     toplevel_fn = config is not None and config._toplevel_fn
 
-    cell_data: list[Union[CellImpl, tuple[str, CellConfig]]] = []
+    # Update old internal cell names to the new ones
+    for idx, name in enumerate(names):
+        if name == "__":
+            names[idx] = DEFAULT_CELL_NAME
+
+    # We require 3 sweeps.
+    #  - One for compilation and import collection
+    #  - One for some basic static determination of top-level functions.
+    #  - And a final sweep for cleaner argument requirements
+    # (since we now know what's top-level).
     defs: set[Name] = set()
+    toplevel_imports: set[Name] = set()
+    used_refs: Optional[set[Name]] = set()
+    import_blocks: list[str] = []
+
+    definition_stubs: list[Optional[CellImpl]] = [None] * len(codes)
+    definitions: list[Optional[str]] = [None] * len(codes)
 
-    cell_id = 0
-    for code, cell_config in zip(codes, cell_configs):
+    cell: Optional[CellImpl]
+    for idx, (code, cell_config) in enumerate(zip(codes, cell_configs)):
         try:
-            cell = compile_cell(code, cell_id=str(cell_id)).configure(
-                cell_config
-            )
+            cell = compile_cell(code, cell_id=str(idx)).configure(cell_config)
             defs |= cell.defs
-            cell_data.append(cell)
+            assert isinstance(used_refs, set)
+            used_refs |= cell.refs
+            if cell.import_workspace.is_import_block:
+                # maybe a bug, but import_workspace.imported_defs does not
+                # contain the information we need.
+                toplevel_imports |= cell.defs
+                if toplevel_fn:
+                    # TODO: Consider fn="imports" for @app.imports?
+                    # Distinguish that something is special about the block
+                    # Also remove the "return" in this case.
+                    definitions[idx] = to_general_functiondef(cell, names[idx])
+                else:
+                    definitions[idx] = to_functiondef(cell, names[idx])
+                import_blocks.append(code.strip())
```

Comment:
If only import blocks are used, then in the below, `foo` won't get saved as a function. I can see this being a bit confusing for users. I'm wondering if imports could be saved top-level even if they weren't in import blocks.

cell:
```python
import random
...
```

Another cell

```python
def foo():
  return random.randint(0, 43)
```

Reply:
Yes they can- but I think restricting to import only blocks makes sense. Consider the following block:

```python {.marimo}
@app.cell
def _(run_button):
    mo.stop(run_button.value)
    import something_very_expensive_with_side_effects
```